### ISSUE TITLE
Helmfile can't access gpg-agent passphrase cache when using secrets plugin

### ISSUE THREAD

**Issue by mherwig** [roboll/helmfile#968]
I use helmfile together with secrets.yaml files in my CI/CD pipeline but it fails to decrypt the secrets because it waits for the passphrase input although I unlocked the gpg-agent earlier and cached the passphrase.

Running `helm secrets dec` works fine but running `helmfile apply` fails or better put: It runs into a timeout because no passphrase it provided. I wonder why this is because I understood that helmfile simply executes `helm secrets dec`.

Currently I'm using a shell script where I do the following:
1. configure gpg-agent to cache for a longer time than the default (ttl settings)
2. import gpg private key
3. encrypt a sample text
4. decrypt the sample text while providing the passphrase
5. installing the chart with `helmfile apply`

Step 5. fails. My current workaround is to run helm secrets dec secrets/secrets.yaml and then refer to secrets/secrets.yaml.dec under the values key in my helmfile.yaml

Do you have an idea why helmfile is not able to use the gpg-agent cache when decvrypting? Seems very odd to me.

Thanks!

**Comment by mumoshu** [comment]
@mherwig Hey! Thanks for trying Helmfile.

I'm not familiar with gpg-agent, but what would be needed for helmfile to interoperate correctly with gpg-agent?

Like, for ssh-agent the process would need a correct `$SSH_AUTH_SOCK` to determine which ssh-agent to connect to. Do we need such envvar, for example?

**Comment by mumoshu** [comment]
```
$ gpg-agent --daemon
```
```
$ ls ~/.gnupg | grep agent
S.gpg-agent
S.gpg-agent.browser
S.gpg-agent.extra
S.gpg-agent.ssh
```

So those `S.gpg-agent.ssh` files seem like what I was asking for.

And it should not depend on any environment variables to discover the `S.gpg-agent` files, as I'm not giving it to my shell.

Hmm, then why it is unable to discover it when run via helmfile and helm-secrets? 🤔

**Comment by mherwig** [comment]
Running  `helm secrets`  does work as expected. As far as I understand the helm secrets plugin, it's just a shell script invoking sops. So all environment variables defined will be accisble by sops.

But helmfile is a go application that runs os.exec, right? I tried the following:

*envtest.go:*
```
package main

import (
	"os"
	"os/exec"
	"log"
)

func main() {
	cmd := exec.Command("bash", "./run_script.sh")
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	err := cmd.Run()

	if err != nil {
		log.Fatalf("cmd.Run() failed with %s\n", err)
	}
}
```

*run_script.sh:*
```
#!/bin/bash

echo $SOME_ENV_VAR
```
If you set `SOME_ENV_VAR`='hello world'`, then run ./gotest, the result will be empty.

So my assumption is, if you set `GPG_TTY=$(tty)`, `GPG_TTY` will be empty for the command you execute via os.exec.

**Comment by mumoshu** [comment]
@mherwig Thanks for your support!

Helmfile does inherit envvars from it to sub-processes like `helm-secrets` here: https://github.com/roboll/helmfile/blob/2f9f52033cc0f0c3cd79ea8435d3aa80e95628fc/pkg/helmexec/runner.go#L35

Perhaps it isn't working as expected, or there's anything that relates to the issue?

I'll keep thinking. I'd appreciate it if you could share your insights like the requirements on `GPG_TTY`.

**Comment by mumoshu** [comment]
@mherwig Btw, do you, by any chance, use helm v2 and have `tillerless: true` set?

We do wrap calls to `helm-secrets` with `helm-tiller` when it is set:
https://github.com/roboll/helmfile/blob/06a3b558e06319489a67ecedc370a8c68ee03f31/pkg/helmexec/exec.go#L192

**Comment by mherwig** [comment]
> @mherwig Btw, do you, by any chance, use helm v2 and have `tillerless: true` set?
> 
> We do wrap calls to `helm-secrets` with `helm-tiller` when it is set:
> https://github.com/roboll/helmfile/blob/06a3b558e06319489a67ecedc370a8c68ee03f31/pkg/helmexec/exec.go#L192

Yes, exactly currently I'm using helm v2.15.2 with `tillerless: true`. Sorry, that I forget to mention.

Running `helm tiller run helm secrets dec secrets.yaml` just fine.
